### ISSUE TITLE
Interval casts raise a ParseException

### ISSUE THREAD

**Issue by rbergm** [klahnakoski/mo-sql-parsing#140]
Parsing queries that contain an interval cast (which is valid in PostgreSQL) raises a `ParseException`:

```python
>>> parse("SELECT '2 months'::interval;")
ParseException: Expecting {integer} | {int} | {int32} | {int64} | {integer} | {int} | {int32} | {int64}, found "interval"
```

The same applies to the explicit cast syntax:

```python
>>> parse("SELECT CAST('2 months' AS INTERVAL);")
ParseException: Expecting {StringEnd} | {from} | {order by} | {offset} | {fetch} | {limit} | {for}, found "('2 months"
```

However, for other types such as date, both variants work fine:

```python
>>> parse("SELECT '2022/11/18'::date;")
{'select': {'value': {'cast': [{'literal': '2022/11/18'}, {'date': {}}]}}}
>>> parse("SELECT CAST('2022/11/18' AS DATE);")
{'select': {'value': {'cast': [{'literal': '2022/11/18'}, {'date': {}}]}}}
```

By the way: I am currently working a lot with queries from the Stack Overflow Benchmark (see [Paper](https://rmarcus.info/stack.html) and [Kaggle](https://www.kaggle.com/datasets/stackoverflow/stackoverflow)) which consists of >5000 machine-generated, as well as hand-crafted queries. This benchmark showed most of the issues that I am reporting. If you are interested, I would be happy to try to integrate the queries as regression tests into your test suite.

**Comment by klahnakoski** [comment]
Thank you for pointing out a source of queries that I can test against.

**Comment by klahnakoski** [comment]
https://github.com/klahnakoski/mo-sql-parsing/commit/7f4218ebd0096b190eb3a7c5178c09e04f2ae9c5

**Comment by klahnakoski** [comment]
https://pypi.org/project/mo-sql-parsing/9.262.22323/

**Comment by klahnakoski** [comment]
leaving open to look at tests, later.

**Comment by rbergm** [comment]
Thank you very much! I was just able to successfully parse and format all 6191 queries of the Stack benchmark.
Due to the size of this benchmark, I was not able to verify, whether the formatted queries indeed match their original counterpart or, whether the parsed structure matches the expected structure. However, all previous attempts failed due to `ParseExceptions`, so I am confident that the `interval` casts where indeed the last missing ingredient to analyzing the benchmark completely.

**Comment by klahnakoski** [comment]
I have also run the tests, and confirmed they do not raise exceptions.  You are correct to point out that we still do not know if the result is correct.  I wonder if running them through the formatter would raise confidence.  Although, the formatter is behind the parser and it could be a lot of work, and still not be certain.

Keeping this open to explore more.

**Comment by rbergm** [comment]
Comparing the parsed+formatted query to the original query actually was my first attempt at regression tests. The main issue I came across here is that the source queries are formatted somewhat arbitrarily and inconsistently. For example, they will sometimes change casing, e.g. `SELECT * FROM foo as f`, contain different amounts of whitespace, e.g. `SELECT * FROM foo,bar, baz`, or feature comments within the query, e.g.

```sql
SELECT *
FROM foo

-- this is very important
WHERE ...
```

This makes comparing the strings themselves very hard (would we need a parser to test the parser?) and in the end I strayed away from this direction.

Another approach to "proper" testing could be to compare the query results - i.e. run the original query and the parsed+formatted query against the same data set and make sure that their result sets are equal. The issue here is the setup of the database. The Stack benchmark is very large (i.e. > 100 GB) and this also makes it much harder to automate such tests. Not to mention that running several thousand queries against this database will probably also take a considerable amount of time.
